narwhal/primary/src/core.rs

```diff
@@ -502,18 +502,29 @@ impl Core {
         notify: Option<oneshot::Sender<DagResult<()>>>,
     ) -> DagResult<()> {
         let digest = certificate.digest();
-        match self.process_certificate_internal(certificate).await {
-            Ok(()) => {
+        if self.certificate_store.read(digest)?.is_some() {
+            trace!("Certificate {digest:?} has already been processed. Skip processing.");
+            self.metrics
+                .duplicate_certificates_processed
+                .with_label_values(&[&certificate.epoch().to_string()])
+                .inc();
+            if let Some(notify) = notify {
+                let _ = notify.send(Ok(())); // no problem if remote side isn't listening
+            }
+            return Ok(());
+        }
+
+        // Sanitize the certificate if we haven't seen it before.
+        if !self.pending_certificates.contains_key(&digest) {
+            if let Err(e) = self.sanitize_certificate(&certificate).await {
```

Comment:
I think we always need to sanitize, or make sure the digest is computed correctly, because a cert can have an existing digest but different content?

Reply:
Ah that's annoying. I assumed digest would include the entire certificate contents but I guess not.

Reply:
I removed this for now but I think it makes this PR less effective unfortunately, I would expect there to be a lot of duplicate certificates processed during catch-up that won't yet be in the certificate store. We can revisit with an alternate approach if it's not adequate without this

Reply:
Sounds good. These related codepaths may see more refactoring, because I think processing certificates (including verifying) can mostly be parallelized in the primary handler or CertificateFetcher.